### PR TITLE
`struct Rav1dData`: Replace `Rav1dRef` with `Option<CArc<[u8]>>`

### PR DESCRIPTION
**Pull request by kkysen**
~~This is still buggy on those 3 tests (`00000063`, `redundant_frame_header`, `issue_48`) failing in #652.  The output video is off by a small number of bits for each of them, but the picture parameters and headers and OBU metadata are all the same.  I've now narrowed down the PR to this smaller one (specifically the 2nd commit, 6afc41b), which still evidently contains the bug.  The divergence in the data happens somewhere in `fn rav1d_decode_tile_sbrow`, but the dataflow gets very confusing and I haven't been able to narrow it down further yet.~~

~~Among the changes in 6afc41b, I also tried not splitting up `Rav1dData` into `CArc<[u8]>` and `Rav1dDataProps` within `fn parse_obus`, but that doesn't change anything.  `fn gen_picture` is also changed in a more straightforward manner that is less concise and efficient as in #652, but the bug is still there.~~

### CODE REVIEW

File: src/lib.rs
Comment (rinon) [inline src/lib.rs:514]: ~~Am I correct in reading this as a deref to the inner slice and calling is_empty() on that? It's hard to figure out types here. If so, how is that ever possible? We don't change the slice size, just the stable_ptr field, right?~~

Ok, this is pretty confusing when we're dealing with slices. Probably worth rethinking some of this or making it more explicit. I see now that we are effectively keeping a slice in the stable_ref and that's what is deref'd, not the actual underlying boxed slice. That's kind of what we want, but it isn't obvious that's what will happen.

File: src/obu.rs
Comment (rinon) [reply src/obu.rs:2209]: I added `assert_eq!(data.len(), pkt_bytelen)` and a lot more tests started failing. `pkt_bytelen` is smaller than `data.len()` in some cases and so this line was shrinking the data slice, in addition to moving the cursor forward.

Comment (rinon) [reply src/obu.rs:2209]: A hacky "shrink" method I added to `CArc<[T]>` and shrinking the slice view to `pkt_bytelen` before moving the start fixed the failing tests.

File: src/lib.rs
Comment (rinon) [reply src/lib.rs:514]: What about making RavidData not a `type` but instead a newtype with a narrower interface? `RavidData` is apparently a ref-counted slice view, which isn't exactly a type Rust already has, so we should be explicit what that type is.

Comment (kkysen) [reply src/lib.rs:514]: Wdym exactly?  It's already a `struct` right now.

Comment (rinon) [reply src/lib.rs:514]: I must have been looking at a different PR, but the point about all the refs and derefs being confusing is still here. It's confusing to have a type that derefs to an inner type except when it's a slice, when it behaves more like a view into that thing. I think I put a comment on the CArc PR that explains that better.

File: src/obu.rs
Comment (kkysen) [reply src/obu.rs:2209]: Thanks for finding the bug!  It's all fixed now, and we don't need a hacky `shrink` method either now that `impl AddAssign` has been replaced with the more flexible `fn slice_in_place`.

File: src/lib.rs
Comment (kkysen) [reply src/lib.rs:514]: I tried to separate the two, but it proved fairly difficult and very unwieldy.  Can I just improve the docs (I might've added some stuff by now)?

Comment (randomPoison) [reply src/lib.rs:514]: @rinon Checking in with you: Is this something to block the PR on? The current version of things seems fairly understandable to me, I'm not sure what has changed between when you last commented and now.

File: src/lib.rs
Comment (randomPoison) [inline src/lib.rs:707]: general nit (not something to necessarily fix here, just a note for the future): I wouldn't bother using the `[markdown link syntax]` in non-doc comments. Keep using \`backticks\` around item names, but the [square brackets] just add noise when they're not actually being used to create a link in the generated docs.

**Review by randomPoison — APPROVED** [review]
Looks good in it's current state but make sure you address Stephen's outstanding comment before merging.

File: src/lib.rs
Comment (kkysen) [reply src/lib.rs:505]: @rinon, does this look okay now?

**Comment by rinon** [comment]
Sorry missed the pings on this, LGTM for now, yeah.

**Pull request merged**
